### PR TITLE
Add Singularity run arguments

### PR DESCRIPTION
**Pull request by davidjurado**
Add singularity feature to allow run arguments, these run arguments can now be defined in the mlcube.yaml file, example:

```yaml
singularity:
        run_args: "-C --writable-tmpfs --net --network=none --nv"
```

In case you are using image conversion from Docker to Singularity (example: using a local Docker tar file), you will need to specify these same arguments under the `docker` section:
```yaml
docker:
        image: my_image:0.0.1
        tar_file: "local_file.tar"
        run_args: "-C --writable-tmpfs --net --network=none --nv"
```

Install the latest version of the Singularity runner to test this new feature:

```bash
virtualenv -p python3 env && source ./env/bin/activate

git clone https://github.com/mlcommons/mlcube && cd ./mlcube

git fetch origin pull/243/head:feature/singularity_run_arguments

git checkout feature/singularity_run_arguments

pip install semver spython && pip install ./mlcube

pip install --no-deps --force-reinstall ./runners/mlcube_singularity
```

### CODE REVIEW

File: runners/mlcube_singularity/mlcube_singularity/singularity_run.py [hunk 1]
```diff
@@ -1,85 +1,220 @@
...
+                    )
+            except:
+                logger.warning(
+                    "SingularityRun can't get singularity version (do you have singularity installed?)."
+                )
+
+            build_file = "docker://" + d_cfg["image"]
+            if "tar_file" in d_cfg:
+                build_file = "docker-archive:" + d_cfg["tar_file"]
+            run_args = d_cfg["run_args"]
```
Comment (mzenk) [inline runners/mlcube_singularity/mlcube_singularity/singularity_run.py:92]:
```suggestion
            run_args = d_cfg.get("run_args", "")
```
I get an error when this argument is not present. I think something like my suggestion should fix this, but haven't tested it

**Review by mzenk — CHANGES_REQUESTED** [review]
There seems to be a problem when `run_args` is not present. I suggested a fix, but didn't look at the rest of the code

**Comment by davidjurado** [comment]
> @davidjurado We're experiencing problems with mlcube since the last commit ([fc31172](https://github.com/mlcommons/mlcube/commit/fc3117292058fd6d91f184ba1be264fa1d1733d9)). When running an mlcube for the FeTS challenge, I get the error:
> 
> ```
> 2022-06-02 14:22:13 e230-pc10 mlcube.shell[23698] ERROR Command='d o c k e r   i n s p e c t   - - t y p e = i m a g e   d o c k e r . s y n a p s e . o r g / s y n 3 1 4 3 7 2 9 3 / f e t s 2 2 _ d a t a - p r e p   >   / d e v / n u l l   2 > & 1', status=512, exit_status=2, on_error=None
> 2022-06-02 14:22:13 e230-pc10 mlcube_docker.docker_run[23698] WARNING Docker image (docker.synapse.org/syn31437293/fets22_data-prep) does not exist or build strategy is 'always'. Will run 'configure' phase.
> Traceback (most recent call last):
>   File "/home/m167k/miniconda3/bin/mlcube", line 8, in <module>
>     sys.exit(cli())
>   File "/home/m167k/miniconda3/lib/python3.7/site-packages/click/core.py", line 829, in __call__
>     return self.main(*args, **kwargs)
>   File "/home/m167k/miniconda3/lib/python3.7/site-packages/click/core.py", line 782, in main
>     rv = self.invoke(ctx)
>   File "/home/m167k/miniconda3/lib/python3.7/site-packages/click/core.py", line 1259, in invoke
>     return _process_result(sub_ctx.command.invoke(sub_ctx))
>   File "/home/m167k/miniconda3/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
>     return ctx.invoke(self.callback, **ctx.params)
>   File "/home/m167k/miniconda3/lib/python3.7/site-packages/click/core.py", line 610, in invoke
>     return callback(*args, **kwargs)
>   File "/home/m167k/miniconda3/lib/python3.7/site-packages/click/decorators.py", line 21, in new_func
>     return f(get_current_context(), *args, **kwargs)
>   File "/home/m167k/miniconda3/lib/python3.7/site-packages/mlcube/__main__.py", line 194, in run
>     docker_runner.run()
>   File "/home/m167k/miniconda3/lib/python3.7/site-packages/mlcube_docker/docker_run.py", line 119, in run
>     self.configure()
>   File "/home/m167k/miniconda3/lib/python3.7/site-packages/mlcube_docker/docker_run.py", line 98, in configure
>     Shell.run(docker, 'pull', image)
> TypeError: run() takes from 1 to 2 positional arguments but 3 were given
> ```
> 
> This setup is not trivial to reproduce, but note that the Shell.run command is different than what this branch has. To reproduce this more easily, I also ran the hello-world example from mlcube_examples (master branch) with `mlcube run --mlcube ~/git_repos/mlcube_examples/hello_world/mlcube.yaml --platform=singularity --task=hello` and got this:
> 
> ```
> 022-06-02 14:57:50 e230-pc10 mlcube_singularity.singularity_run[30978] WARNING SingularityRun singularity configuration not found in MLCube file (singularity=None).
> Traceback (most recent call last):
>   File "/home/m167k/tmp_workspace_badmlcube/env/bin/mlcube", line 8, in <module>
>     sys.exit(cli())
>   File "/home/m167k/tmp_workspace_badmlcube/env/lib/python3.10/site-packages/click/core.py", line 829, in __call__
>     return self.main(*args, **kwargs)
>   File "/home/m167k/tmp_workspace_badmlcube/env/lib/python3.10/site-packages/click/core.py", line 782, in main
>     rv = self.invoke(ctx)
>   File "/home/m167k/tmp_workspace_badmlcube/env/lib/python3.10/site-packages/click/core.py", line 1259, in invoke
>     return _process_result(sub_ctx.command.invoke(sub_ctx))
>   File "/home/m167k/tmp_workspace_badmlcube/env/lib/python3.10/site-packages/click/core.py", line 1066, in invoke
>     return ctx.invoke(self.callback, **ctx.params)
>   File "/home/m167k/tmp_workspace_badmlcube/env/lib/python3.10/site-packages/click/core.py", line 610, in invoke
>     return callback(*args, **kwargs)
>   File "/home/m167k/tmp_workspace_badmlcube/env/lib/python3.10/site-packages/click/decorators.py", line 21, in new_func
>     return f(get_current_context(), *args, **kwargs)
>   File "/home/m167k/tmp_workspace_badmlcube/env/lib/python3.10/site-packages/mlcube/__main__.py", line 194, in run
>     docker_runner.run()
>   File "/home/m167k/tmp_workspace_badmlcube/env/lib/python3.10/site-packages/mlcube_singularity/singularity_run.py", line 201, in run
>     self.configure()
>   File "/home/m167k/tmp_workspace_badmlcube/env/lib/python3.10/site-packages/mlcube_singularity/singularity_run.py", line 185, in configure
>     Shell.run(
> TypeError: Shell.run() got an unexpected keyword argument 'die_on_error'
> ```

Hello @mzenk, I just tested another implementation and you are right there was a problem with the Shell.run command, so I applied the fix, for the hello world example I'm now getting a different error but it's related to the example implementation. Please take a look and let me know if there is any other problem, thanks!

**Comment by mzenk** [comment]
@davidjurado After the cubes kept failing with the same error, I figured out what is the problem. Actually it wasn't about the singularity runner code here, but about the installation instructions. Specifically, you switched mlcube version requirement for the runner to 0.0.5 at some point, but did not update the line
```
pip install semver spython && pip install ./mlcube
```
in the instructions. This installs mlcube from the local git repo, which unfortunately is NOT identical to 0.0.5. The Shell.run signature is different in your branch than in 0.0.5. Installing mlcube from pypi instead solved the problem for me, so
```
pip install semver spython mlcube==0.0.5
```

**Pull request merged**